vmdb/app/services/charts_layout_service.rb

```diff
@@ -0,0 +1,61 @@
+class ChartsLayoutService
+  def initialize(perf_record, charts_folder, layout, fname = nil)
+    # DB record that has the charts, e.g. Host, EmsCluster, etc.
+    @perf_record   = perf_record
+    # Main folder of charts layouts
+    @charts_folder = charts_folder
+    # Name of the layout dir or file
+    @layout        = layout
+    # Optional name of the file in the layout directory, usually it's name of the base class
+    @fname         = fname
+  end
+
+  def layout
+    charts = build_charts
+    charts.delete_if do |c|
+      c.is_a?(Hash) && c[:applies_to_method] && @perf_record &&
+          @perf_record.respond_to?(c[:applies_to_method]) &&
+          !@perf_record.send(c[:applies_to_method].to_sym)
+    end
+  end
+
+  private
+
+  def build_charts
+    if @fname
+      if class_specific_layout?
+        # First check if there is file for specific class of the @perf_record
+        load_yaml(class_specific_path)
+      else
+        # Fallback to fname, which is usually base class of the @perf_record
+        load_yaml(fname_path)
+      end
+    else
+      load_yaml(layout_path)
+    end
+  end
```

Comment:
still too much duplication I think

what about something like this:

```
def find_chart_path
  if @fname
    path = make_filename(@layout, @perf_record.class.name)
    return path if File.exists?(path)
    make_filename(@layout, @fname)
  else
    make_filename(@layout)
  end
end

def build_charts
   YAML::load(File.open(find_chart_path))
end

def make_filename(*base)
    File.join(@charts_folder, *base) + '.yaml'
end
```  

I am using your logic above, just simplifying your code although I don't see why you need the @fname when using the `@perf_record.class.name`